Bug: returning reference to local temporary object

Clang reports the following warning:

```
../single_include/nlohmann/json.hpp:18265:20: error: returning reference to local temporary object [-Werror,-Wreturn-stack-address]
            return *ptr;
                   ^~~~
```

Context:

```cpp
template<typename ReferenceType, typename ThisType>
static ReferenceType get_ref_impl(ThisType& obj)
{
    // delegate the call to get_ptr<>()
    auto ptr = obj.template get_ptr<typename std::add_pointer<ReferenceType>::type>();

    if (JSON_HEDLEY_LIKELY(ptr != nullptr))
    {
        return *ptr;          // <----- warning
    }

    JSON_THROW(type_error::create(303, "incompatible ReferenceType for get_ref, actual type is " + std::string(obj.type_name())));
}
```

More context:

```
In file included from ../test/src/unit-udt.cpp:32:
../include/nlohmann/json.hpp:2779:20: warning: returning reference to local temporary object [-Wreturn-stack-address]
            return *ptr;
                   ^~~~
../include/nlohmann/json.hpp:3136:16: note: in instantiation of function template specialization 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::get_ref_impl<const nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::internal_binary_t &, const nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > > >' requested here
        return get_ref_impl<ReferenceType>(*this);
               ^
../include/nlohmann/json.hpp:1490:77: note: in instantiation of function template specialization 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::get_ref<const nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::internal_binary_t &, 0>' requested here
                JSONSerializer<other_binary_t>::to_json(*this, val.template get_ref<const other_binary_t&>());
                                                                            ^
../test/src/unit-udt.cpp:714:26: note: in instantiation of function template specialization 'nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, another_adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >::basic_json<nlohmann::basic_json<std::map, std::vector, std::__1::basic_string<char>, bool, long long, unsigned long long, double, std::allocator, adl_serializer, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >, 0>' requested here
        custom_json cj = j;
                         ^
```

--- a/include/nlohmann/json.hpp
+++ b/include/nlohmann/json.hpp
@@ -2757,6 +2757,18 @@ class basic_json
         return is_binary() ? m_value.binary : nullptr;
     }
 
+    /// get a pointer to the value (binary)
+    internal_binary_t* get_impl_ptr(internal_binary_t* /*unused*/) noexcept
+    {
+        return is_binary() ? m_value.binary : nullptr;
+    }
+
+    /// get a pointer to the value (binary)
+    constexpr const internal_binary_t* get_impl_ptr(const internal_binary_t* /*unused*/) const noexcept
+    {
+        return is_binary() ? m_value.binary : nullptr;
+    }
+
     /*!
     @brief helper function to implement get_ref()
 

--- a/single_include/nlohmann/json.hpp
+++ b/single_include/nlohmann/json.hpp
@@ -18243,6 +18243,18 @@ class basic_json
         return is_binary() ? m_value.binary : nullptr;
     }
 
+    /// get a pointer to the value (binary)
+    internal_binary_t* get_impl_ptr(internal_binary_t* /*unused*/) noexcept
+    {
+        return is_binary() ? m_value.binary : nullptr;
+    }
+
+    /// get a pointer to the value (binary)
+    constexpr const internal_binary_t* get_impl_ptr(const internal_binary_t* /*unused*/) const noexcept
+    {
+        return is_binary() ? m_value.binary : nullptr;
+    }
+
     /*!
     @brief helper function to implement get_ref()
 

--- a/test/src/unit-constructor1.cpp
+++ b/test/src/unit-constructor1.cpp
@@ -826,7 +826,7 @@ TEST_CASE("constructors")
 
             // check that NaN is serialized to null
             CHECK(j.dump() == "null");
-	}
+        }
 
         SECTION("infinity")
         {

